q2cli/util.py

```diff
@@ -234,3 +275,119 @@ def get_plugin_manager():
             return pm
 
         return qiime2.sdk.PluginManager()
+
+
+def load_metadata(fp):
+    """ Turns a filepath into metadata if the path is either to metadata or an
+    artifact that can be viewed as metadata
+    """
+    import sys
+    import qiime2
+
+    try:
+        artifact = get_input(fp)
+    except ControlFlowException:
+        try:
+            return qiime2.Metadata.load(fp)
+        except Exception as e:
+            # Handle the case where we try to load a raw .qza as Metadata
+            # because we were missing the pugin(s) necessary to load it as an
+            # Artifact
+            if "'utf-8' codec can't decode byte" in str(e) and \
+                    qiime2.sdk.result.Result.peek(fp):
+                e = ValueError("It looks like you have an Artifact but are "
+                               "missing the plugin(s) necessary to load it.")
+
+            header = ("There was an issue with loading the file %s as "
+                      "metadata:" % fp)
+            tb = 'stderr' if '--verbose' in sys.argv else None
+            exit_with_error(e, header=header, traceback=tb)
+
+    if isinstance(artifact, qiime2.Visualization):
+        raise Exception(
+            f'Visualizations cannot be viewed as QIIME 2 metadata:\n{fp}')
+    elif artifact.has_metadata():
+        try:
+            metadata = artifact.view(qiime2.Metadata)
+        except Exception as e:
+            header = ("There was an issue with viewing the artifact "
+                      "%s as QIIME 2 Metadata:" % fp)
+            tb = 'stderr' if '--verbose' in sys.argv else None
+            exit_with_error(e, header=header, traceback=tb)
+    else:
+        raise Exception("Artifacts with type %r cannot be viewed as"
+                        " QIIME 2 metadata:\n%r" % (artifact.type, fp))
+
+    return metadata
+
+
+def get_input(fp):
+    """ Gets a Result from a filepath if possible
+    """
+    import tempfile
+    import qiime2
+    import qiime2.sdk
+
+    get_plugin_manager()
+    try:
```

Comment:
I think there may be one too many try's here, or at least I'm struggling to follow what's expected to happen in different situations.

I think this may need some refactoring, but I'm not really sure how to go about it either. It seems like there must be phases we can distill?

Reply:
The try here is handling all the errors with loading an Artifact that we understand. If we get an issue that we don't understand we raise a ControlFlowException that can be handled by the call site. The reason the call sites also have try's is so _convert_input can pretty print the failure if there is an exception and so load_metadata can try to load the file as metadata instead of as an artifact if the error it got was a ControlFlowException (any error loading the artifact that we don't understand).